lib/OpenQA/Utils.pm

```diff
@@ -224,71 +224,68 @@ sub needle_info {
 }
 
 # logging helpers
-sub log_debug {
-    my ($msg, $channel) = @_;
 
-    if ($channel && $channels{$channel}) {
-        $channels{$channel}->debug($msg);
-    }
-    elsif ($app && $app->log) {
-        $app->log->debug($msg);
-    }
-    else {
-        STDOUT->printflush("[DEBUG] $msg\n");
-    }
+sub log_debug {
+    _log_msg('debug', @_);
 }
 
 sub log_info {
-    my ($msg, $channel) = @_;
-    if ($channel && $channels{$channel}) {
-        $channels{$channel}->info($msg);
-    }
-    elsif ($app && $app->log) {
-        $app->log->info($msg);
-    }
-    else {
-        STDOUT->printflush("[INFO] $msg\n");
-    }
+    _log_msg('info', @_);
 }
 
 sub log_warning {
-    my ($msg, $channel) = @_;
-    if ($channel && $channels{$channel}) {
-        $channels{$channel}->warn($msg);
-    }
-    elsif ($app && $app->log) {
-        $app->log->warn($msg);
-    }
-    else {
-        STDERR->printflush("[WARN] $msg\n");
-    }
+    _log_msg('warn', @_);
 }
 
 sub log_error {
-    my ($msg, $channel) = @_;
-    if ($channel && $channels{$channel}) {
-        $channels{$channel}->error($msg);
-    }
-    elsif ($app && $app->log) {
-        $app->log->error($msg);
-    }
-    else {
-        STDERR->printflush("[ERROR] $msg\n");
-    }
+    _log_msg('error', @_);
 }
 
 sub log_fatal {
-    my ($msg, $channel) = @_;
+    _log_msg('fatal', @_);
+    die $_[0];
+}
+
+sub _log_msg {
+    my ($level, $msg, %options) = @_;
+    my $code = <<"END";
+    my (\$msg, \%options) = \@_;
+    my \$log_default = \$options{default};
+
+    if(\$options{channels}) {
+        if(ref(\$options{channels}) eq 'ARRAY') {
+            for my \$channel (\@{\$options{channels}}) {
+                _log (\$level, \$msg,\$channel);
+            }
+        } else {
+            _log(\$level, \$msg, \$options{channels});
+        }
+    } else {
+        \$log_default = 1;
+    }
+    _log(\$level, \$msg) if \$log_default;
+END
+    my $sub = sub { eval($code) };
+    eval { $sub->($msg, %options) };
+}
+
+
+sub _log {
+    my ($level, $msg, $channel) = @_;
     if ($channel && $channels{$channel}) {
-        $channels{$channel}->fatal($msg);
+        eval("\$channels{$channel}->$level(\$msg);");
```

Comment:
`$channels{$channel}->$level($msg);` should be perfectly fine without the `eval`.

Reply:
You're right :-)